### PR TITLE
Add `--allow-custom-scopes` flag to skip scope validation.

### PR DESCRIPTION
**Pull request by Haard30**
This PR adds a `--allow-custom-scopes` flag to skip scope validation for PAT. This allows addition undocumented ADO scopes.

Testing:
1. PAT with just 1 invalid scope.
Earlier
image

With `--allow-custom-scopes` flag

image

```
microsoft-authentication-cli\dist on  user/haashah/add-no-validate-flag [!]
❯ .\azureauth.exe ado pat  --organization "office" --scope "lolwut" --display-name "Test PAT" --prompt-hint "Creating custom PAT test" --allow-custom-scopes
Skipping scopes validation.
<pat>
```

2. PAT with 1 known and 1 invalid scope
```
❯ .\azureauth.exe ado pat  --organization "office" --scope "lolwut" --scope "vso.code" --display-name "Test PAT 2" --prompt-hint "Creating PAT for testing" --allow-custom-scopes
Skipping scopes validation.
One or more errors occurred. (Failed to create PAT: InvalidScope)
An unexpected error ocurred. Re-run with --verbosity=debug for more info.
```
3. PAT with valid known ADO scope and valid unknown ADO scope.

```
❯ .\azureauth.exe ado pat  --organization "office" --scope "vso.governance" --scope "vso.code" --display-name "Non-public valid scope" --prompt-hint "Creating PAT for testing" --allow-custom-scopes
Skipping scopes validation.
<pat>
```

4. Without `--allow-custom-scopes` flag
```
❯  .\azureauth.exe ado pat  --organization "office" --scope "vso.governance" --scope "vso.code" --display-name "Non-public valid scope" --prompt-hint "Creating PAT for testing"
vso.governance is not a known Azure DevOps PAT scope.
Consult https://aka.ms/azure-devops-pat-scopes for a list of known scopes.
Use --allow-custom-scopes to create a PAT with custom scopes.
```

### CODE REVIEW

File: src/AzureAuth/Commands/Ado/CommandPat.cs [hunk 1]
```diff
@@ -36,6 +36,9 @@ public class CommandPat
         private const string ScopeOption = "--scope";
         private const string ScopeHelp = "A token scope for accessing Azure DevOps resources. Repeated invocations allowed.";
 
+        private const string NoValidationFlag = "--no-validate";
+        private const string NoValidationHelp = "Flag to skip scope validation. Allows addition of non-public ADO scopes.";
```
Comment (kyle-rader-msft) [inline src/AzureAuth/Commands/Ado/CommandPat.cs:40]: Consider shortening to 
```
"Skip scopes validation"
```

Comment (Haard30) [reply src/AzureAuth/Commands/Ado/CommandPat.cs]: Makes sense. Done!

File: src/AzureAuth/Commands/Ado/CommandPat.cs [hunk 2]
```diff
@@ -168,15 +174,22 @@ private bool ValidOptions(ILogger logger)
             }
             else
             {
-                var invalidScopes = AdoPat.Scopes.Validate(this.Scopes);
-                if (!invalidScopes.IsEmpty)
+                if(NoScopeValidation)
+                {
+                    logger.LogWarning($"Received {NoValidationFlag} flag. AzureAuth will skip validation of scopes.");
```
Comment (kyle-rader-msft) [inline src/AzureAuth/Commands/Ado/CommandPat.cs:179]: Could also simplify this message to "Skipping scopes validation"

File: src/AzureAuth/Commands/Ado/CommandPat.cs [hunk 3]
```diff
@@ -168,15 +174,22 @@ private bool ValidOptions(ILogger logger)
...
-                    foreach (var scope in invalidScopes)
+                    var invalidScopes = AdoPat.Scopes.Validate(this.Scopes);
+                    if (!invalidScopes.IsEmpty)
                     {
-                        logger.LogError($"{scope} is not a valid Azure DevOps PAT scope.");
+                        foreach (var scope in invalidScopes)
+                        {
+                            logger.LogError($"{scope} is not a valid Azure DevOps PAT scope.");
+                        }
+                        logger.LogError($"Consult {AdoPat.Constants.PatListURL} for a list of valid scopes.");
```
Comment (shalinikhare27) [inline src/AzureAuth/Commands/Ado/CommandPat.cs:190]: We use the word `consult` but I guess all we recommend is that this link is your source of truth and if you still want to use custom scope then just use it with the `--no-validate` flag?

**Review by mvanchaa — APPROVED** [review]

File: src/AzureAuth/Commands/Ado/CommandPat.cs (hunk 2 above)
Comment (kyle-rader-msft) [reply src/AzureAuth/Commands/Ado/CommandPat.cs:179]: Consider moving this nested check into the outer if/else block so you have 3 main branches:
1. no scopes, print error, return false
2. there are scopes, but skipValidation is passed, warn, do nothing else,
3. do scope validation

**Review by mijpeterson — APPROVED** [review]

File: src/AzureAuth/Commands/Ado/CommandPat.cs [hunk 4]
```diff
@@ -168,15 +174,22 @@ private bool ValidOptions(ILogger logger)
...
+                    logger.LogWarning($"Received {NoValidationFlag} flag. AzureAuth will skip validation of scopes.");
+                }
+                else
                 {
-                    foreach (var scope in invalidScopes)
+                    var invalidScopes = AdoPat.Scopes.Validate(this.Scopes);
+                    if (!invalidScopes.IsEmpty)
                     {
-                        logger.LogError($"{scope} is not a valid Azure DevOps PAT scope.");
+                        foreach (var scope in invalidScopes)
```
Comment (shalinikhare27) [inline src/AzureAuth/Commands/Ado/CommandPat.cs:186]: The error message does indicate that the scope is not valid and so should look at official Microsoft ADO scope list. However, we should also tell users that if they are still wanting to use that scope, then they can use it with the flag. Basically how will the user discover this flag exists if they really want to use the scope.

**Comment by reillysiemens** [comment]
Consider renaming this option to `--allow-custom-scopes` as that might be a little bit more self-explanatory.

File: src/AzureAuth/Commands/Ado/CommandPat.cs (hunk 4 above)
Comment (shalinikhare27) [reply src/AzureAuth/Commands/Ado/CommandPat.cs:186]: Apart of running `azureauth ado pat --help` I guess.

Comment (Haard30) [reply src/AzureAuth/Commands/Ado/CommandPat.cs]: Added that to the message

File: src/AzureAuth/Commands/Ado/CommandPat.cs (hunk 3 above)
Comment (Haard30) [reply src/AzureAuth/Commands/Ado/CommandPat.cs]: Updated to `known/unknown`

File: src/AzureAuth/Commands/Ado/CommandPat.cs (hunk 4 above)
Comment (shalinikhare27) [reply src/AzureAuth/Commands/Ado/CommandPat.cs]: Thanks @Haard30  for adding. Makes the discoverability easier/better, so better user experience on next steps.

File: src/AzureAuth/Commands/Ado/CommandPat.cs (hunk 2 above)
Comment (Haard30) [reply src/AzureAuth/Commands/Ado/CommandPat.cs]: Done! Thanks.

**Review by shalinikhare27 — APPROVED** [review]

**Review by reillysiemens — APPROVED** [review]

**Review by mvanchaa — APPROVED** [review]

**Pull request merged**